alphagov/paas-usage-events-collector
# [#152241494] Billing API

## What

We are extending the event collector application to give us a "Billing API" that allows us to query resource usage and calculate costs for that usage. It may make sense to rename this repository "paas-billing" in the future, and use it to contain those items that are dependent on the billing database.

* A `MATERIALIZED VIEW` is created in the database that gets updated at a set interval (currently 1hr) which normalizes the raw event data into rows of `resource_guid`, `time period`, `pricing_plan_id` , then joins the rows with data from a `pricing_plans` table that contains the information required to calculate the prices. The view caches and indexes data in the following form so that we can perform queries against it efficiently. 

| ... | duration | guid | org | space | pricing_plan_id | price | ... |
| --- | --- | --- | --- | --- | --- | --- | --- |
| ... | `2017-11-01 08:14` to `2017-11-19 08:14` | {thing1} | {org1} | {space1} | {planA} | £x | ... |
| ... | `2017-11-01 09:10` to `2017-11-19 10:10` | {thing2} | {org2} | {space1} | {planB} | £x | ... |


* A `pricing_plans` table is added to the database that contains the formulas required to calculate costs for each resource. For example a pricing_plan row might look like:


| id | name | valid_from | plan_guid | formula |
| --- | --- | --- | --- | --- |
| 2 | compute instance | 2017-11-01 | {plan_guid} | `($memory_in_mb / 1000) * ($time_in_seconds)` |
| 3 | tiny postgres | 2017-11-01 | {plan_guid} | `1 * $time_in_seconds` |

* Pricing plans can change over time so they have a valid_from field. The view calculation handles splitting usage over the valid ranges.

* A REST/JSON API exposes aggregated data at several levels. Only guid details are returned in the data at the moment. If you want names you would need to call out the cf:
    - `/organisations` list totals for all orgs
    - `/organisations/:org_guid` list total for a single org
    - `/organisations/:org_guid/spaces` list totals for each space in an org
    - `/organisations/:org_guid/resources` list totals for each resource in org (a resource is a single thing - like an "app" or a "service")
    - `/spaces` list totals for all spaces
    - `/spaces/:space_guid` list total for a single space
    - `/spaces/:space_guid/resources` list totals for each resource in space
    - `/resources` list totals for all resources
    - `/resources/:resource_guid` list total for a single resource
    - `/events` [experimental] list all events ("events" are all the start/stop points with calculated billing, unlike "resources" which are aggregate totals for each item over a range). Events would allow you to see _when_ something happens
    - `/resources/:resource_guid/events` as above but for a single resource
    - `/pricing_plans` fetch the pricing plans

* A (throwaway) example HTML rendering of an aggregated report can be found at `/` you will be prompted to login via UAA. This is meant purely as an illustration of what is possible for now.

## How to review

* Deploy to dev using https://github.com/alphagov/paas-cf/pull/1116
* Go start/stop some apps/services to generate some events in your dev environment (ie. restart the healthcheck apps, create a postgres service etc
* Visit https://paas-usage-events-collector.$DEPLOY_ENV.dev.cloudpipelineapps.digital/
* Login as an admin when prompted and authorize the app 
* Visit /pricing_plans and create a pricing plan for the compute plan with guid: `f4d4b95a-f55e-4593-8d54-3364c25798c4`
* Visit /pricing_plans and press the "Seed pricing plans" button to create initial pricing plans for all known services.
* Visit `/` to see an HTML report of some pricing
* Query the Billing API for some results

## Who can review

Not @chrisfarms, @46bit, @camelpunch, @samcrang
